### PR TITLE
feat: Instrument AST in debug mode to track variable state

### PR DESCRIPTION
**Pull request by nthiad**
# Description

This patch builds on existing debugger work to add support for a `vars` command to inspect variable values as the debugger steps through instructions. This is accomplished by instrumenting the AST after parsing with foreign calls to debug methods for tracking assignments and drops. There is an additional pass during monomorphization that collects type info.

The instrumentation approach ensures that variables aren't optimized away so that users can step through program logic with variable names as they appear in the original source code, not only with witness maps and register values.

This patch isn't yet ready to be merged in but is available for feedback.

Here's what it looks like to use, demonstrating a shadowed "a" variable. (Ignore the source pointer bug)

```
At /home/nthia/dev/manas/noir/tooling/nargo_cli/tests/execution_success/arithmetic_binary_operations/src/main.nr:18:9
 12 ...
 13        let b = a - y; // 6 - 4 = 2
 14        let c = b * z; // 2 * 5 = 10
 15        {
 16            let mut a = 3;
 17            a *= 111;
 18 ->         println(f"a={a}");
 19        }
 20        let d = c / a; // 10 / 6 (This uses field inversion, so we test it by multiplying by `a`)
 21        d * a
 22    }
> vars
z:Field=5, x:Field=3, b:Field=2, a:Field=3, a:Field=6, y:Field=4, c:Field=10
```

## Problem\*

This feature is part of #3015 and other related work on the debugger. The instrumentation is also useful for integrating with DAP (sending type and callstack info) and tracking callstack parameters.

Earlier into this feature, I explored mapping variables onto witness map entries and register values, but that approach quickly proved to be much more difficult because optimizations happen at multiple stages of the compiler process that destroy information required to establish mappings, and adding more info would be much more difficult than adding instrumentation instructions. Instrumentation also makes other features easier to add beyond only variable tracking such as tracking call stack parameters.

## Summary\*

* Adds a `vars` command to the debugger
* Instruments code with foreign call wrappers in debug mode to track variable state
* Works with shadowing but does not yet work with callstacks.

## Additional Context

The double-wrapping in the injected oracles for the debug foreign calls is intentional and without this the compiler complains about ``All `oracle` methods should be wrapped in an unconstrained fn``.

This patch is mostly ready but has a few known issues to work on. I will update this list as they are completed:

- [ ] instrumented AST nodes break source offset tracking
- [ ] ensure debugging instrumentation is only turned on in the debugger and can be turned on/off manually
- [ ] member and index assignment
- [ ] display non-scalar data types (structs, tuples, arrays), reading the appropriate number of fields
- [ ] DebugState can probably be removed from the context instance now, if not removed altogether due to the extra pass added to the monomorphize stage

Future work will add tests for various instrumentation cases and will track variables against the callstack (likely with an additional foreign method). Right now the assignment is flat but it does work properly with shadowing.

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

The documentation for this feature will be added along with the rest of the debugger documentation, as mentioned in #2995.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: tooling/nargo/src/lib.rs [hunk 1]
```diff
@@ -49,6 +49,7 @@ pub fn prepare_package(package: &Package, file_reader: Box<FileReader>) -> (Cont
     let mut context = Context::new(fm, graph);
 
     let crate_id = prepare_crate(&mut context, &package.entry_path);
+    context.root_crate_id = crate_id.clone();
```
Comment (TomAFrench) [inline tooling/nargo/src/lib.rs:52]:
```suggestion
    context.root_crate_id = crate_id;
```

This is `Copy` so a clone is unnecessary.

File: tooling/nargo/src/artifacts/debug_vars.rs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    id_to_value: HashMap<u32, Value>, // TODO: something more sophisticated for lexical levels
+    id_to_type: HashMap<u32, u32>,
+    types: HashMap<u32, PrintableType>,
+}
+
+impl DebugVars {
+    pub fn new(vars: &HashMap<u32, String>) -> Self {
+        let mut debug_vars = Self::default();
+        debug_vars.id_to_name = vars.clone();
+        debug_vars
```
Comment (TomAFrench) [inline tooling/nargo/src/artifacts/debug_vars.rs:18]:
```suggestion
        Self {
            id_to_name: vars.clone()
            ..Self::default()
        }
```

File: tooling/nargo/src/artifacts/debug_vars.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    pub fn get_variables<'a>(&'a self) -> Vec<(&'a str, &'a Value, &'a PrintableType)> {
```
Comment (TomAFrench) [inline tooling/nargo/src/artifacts/debug_vars.rs:21]:
```suggestion
    pub fn get_variables(&self) -> Vec<(&str, &Value, &PrintableType)> {
```

Lifetimes aren't necessary here.

File: tooling/nargo/src/artifacts/debug_vars.rs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                self.id_to_name
+                    .get(var_id)
+                    .and_then(|name| self.id_to_value.get(var_id).map(|value| (name, value)))
+                    .and_then(|(name, value)| {
+                        self.id_to_type.get(var_id).map(|type_id| (name, value, type_id))
+                    })
+                    .and_then(|(name, value, type_id)| {
+                        self.types.get(type_id).map(|ptype| (name.as_str(), value, ptype))
+                    })
+            })
```
Comment (TomAFrench) [inline tooling/nargo/src/artifacts/debug_vars.rs:34]:
```suggestion
                let name = self.id_to_name.get(var_id)?;
                let value = self.id_to_value.get(var_id)?;
                let type_id = self.id_to_type.get(var_id)?;
                let printable_type = self.types.get(type_id)?;

                Some((name.as_str(), value, printable_type))
```

We can avoid the usage of `and_then` with `?` which greatly helps readability imo.

File: tooling/nargo/src/artifacts/debug_vars.rs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    pub fn assign_index(&mut self, _var_id: u32, _index: u64, _value: Value) {
+        // TODO
+    }
+
+    pub fn assign_deref(&mut self, _var_id: u32, _value: Value) {
+        // TODO
+    }
+
+    pub fn get<'a>(&'a mut self, var_id: u32) -> Option<&'a Value> {
```
Comment (TomAFrench) [inline tooling/nargo/src/artifacts/debug_vars.rs:68]:
```suggestion
    pub fn get(&mut self, var_id: u32) -> Option<&Value> {
```

File: compiler/noirc_errors/src/debug_info.rs [hunk 6]
```diff
@@ -18,6 +23,8 @@ pub struct DebugInfo {
     /// that they should be serialized to/from strings.
     #[serde_as(as = "BTreeMap<DisplayFromStr, _>")]
     pub locations: BTreeMap<OpcodeLocation, Vec<Location>>,
+    pub variables: HashMap<u32, (String, u32)>, // var_id => (name, type_id)
+    pub types: HashMap<u32, PrintableType>,     // type_id => printable type
```
Comment (TomAFrench) [inline compiler/noirc_errors/src/debug_info.rs:27]: Note to self to look at how this acts when we're not in a debug context. Should we make this optional?

File: tooling/nargo/src/ops/foreign_calls.rs
Comment (TomAFrench) [inline tooling/nargo/src/ops/foreign_calls.rs:1]: I'm considering whether we want to have these foreign calls defined the the debugger crate instead. Ideally the `DefaultForeignCallExecutor` would be as thin as possible.

File: compiler/noirc_errors/src/debug_info.rs [hunk 7]
```diff
@@ -85,4 +99,11 @@ impl DebugInfo {
 
         counted_opcodes
     }
+
+    pub fn get_file_ids(&self) -> Vec<FileId> {
+        self.locations
+            .values()
+            .filter_map(|call_stack| call_stack.last().map(|location| location.file))
```
Comment (TomAFrench) [inline compiler/noirc_errors/src/debug_info.rs:106]: I'm pretty sure we can't have empty callstacks so this filtering may be unnecessary.

File: compiler/noirc_frontend/src/debug/mod.rs [hunk 8]
```diff
@@ -0,0 +1,503 @@
...
+            .collect();
+
+        let set_fn_params = pvars
+            .iter()
+            .map(|(var_id, id, _is_mut)| self.wrap_assign_var(*var_id, id_expr(id)))
+            .collect();
+
+        self.walk_scope(&mut f.body.0);
+
+        // prapend fn params:
```
Comment (TomAFrench) [inline compiler/noirc_frontend/src/debug/mod.rs:60]:
```suggestion
        // prepend fn params:
```

nit

File: tooling/debugger/Cargo.toml
Comment (TomAFrench) [inline tooling/debugger/Cargo.toml:1]: We need to update Cargo.toml to reflect this.

**Review by TomAFrench — COMMENTED** [review]
I'm going to hold off from digging too deep into this as it seems like there's additional changes to be made so just a few small notes.

**Comment by TomAFrench** [comment]
Marking this as draft until the remaining TODOs are addressed.

**Comment by mverzilli** [comment]
@TomAFrench this has now been superceded by https://github.com/noir-lang/noir/pull/4122, which should have also addressed all your comments here.

@ggiraldez mentioning you here in case you weren't aware of this PR.

@nthiad if you see this notification, would you close this PR? :).

**Pull request closed without merging**